spark/src/main/scala/org/apache/spark/sql/delta/util/threads/sparkThreadLocalForwardingThreadPoolExecutor.scala

```diff
@@ -0,0 +1,132 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.ThreadUtils.namedThreadFactory
+
+/**
+ * Implementation of ThreadPoolExecutor that captures the Spark ThreadLocals present at submit time
+ * and inserts them into the thread before executing the provided runnable.
+ */
+class SparkThreadLocalForwardingThreadPoolExecutor(
+    corePoolSize: Int,
+    maximumPoolSize: Int,
+    keepAliveTime: Long,
+    unit: TimeUnit,
+    workQueue: BlockingQueue[Runnable],
+    threadFactory: ThreadFactory,
+    rejectedExecutionHandler: RejectedExecutionHandler = new ThreadPoolExecutor.AbortPolicy)
+  extends ThreadPoolExecutor(
+    corePoolSize, maximumPoolSize, keepAliveTime,
+    unit, workQueue, threadFactory, rejectedExecutionHandler) {
+
+  override def execute(command: Runnable): Unit =
+    super.execute(new SparkThreadLocalCapturingRunnable(command))
+}
+
+object SparkThreadLocalForwardingThreadPoolExecutor {
+  /**
+   * Create a cached thread pool whose max number of threads is `maxThreadNumber`. Thread names
+   * are formatted as prefix-ID, where ID is a unique, sequentially assigned integer.
+   */
+  def newDaemonCachedThreadPool(
+      prefix: String,
+      maxThreadNumber: Int,
+      keepAliveSeconds: Int = 60,
+      queueSize: Int = Integer.MAX_VALUE): ThreadPoolExecutor = {
+    val threadFactory = namedThreadFactory(prefix)
+    val threadPool = new SparkThreadLocalForwardingThreadPoolExecutor(
+      maxThreadNumber, // corePoolSize: the max number of threads to create before queuing the tasks
+      maxThreadNumber, // maximumPoolSize: because we use LinkedBlockingDeque, this one is not used
+      keepAliveSeconds,
+      TimeUnit.SECONDS,
+      new LinkedBlockingQueue[Runnable](queueSize),
+      threadFactory)
+    threadPool.allowCoreThreadTimeOut(true)
+    threadPool
+  }
+}
+
+trait SparkThreadLocalCapturingHelper extends Logging {
+  // At the time of creating this instance we capture the task context and command context.
+  val capturedTaskContext = TaskContext.get()
+  val sparkContext = SparkContext.getActive
+  // Capture an immutable threadsafe snapshot of the current local properties
+  val capturedProperties = sparkContext
+    .map(sc => CapturedSparkThreadLocals.toValuesArray(
+      org.apache.spark.util.Utils.cloneProperties(sc.getLocalProperties)))
+
+  def runWithCaptured[T](body: => T): T = {
+    // Save the previous contexts, overwrite them with the captured contexts, and then restore the
+    // previous when execution completes.
+    // This has the unfortunate side effect of writing nulls to these thread locals if they were
+    // empty beforehand.
+    val previousTaskContext = TaskContext.get()
+    val previousProperties = sparkContext.map(_.getLocalProperties)
+
+    TaskContext.setTaskContext(capturedTaskContext)
+    capturedProperties.foreach { p =>
+      sparkContext.foreach(_.setLocalProperties(CapturedSparkThreadLocals.toProperties(p)))
+    }
```

Comment:
```suggestion
    for {
      p <- capturedProperties
      sc <- sparkContext
    } sc.setLocalProperties(CapturedSparkThreadLocals.toProperties(p))
```
Isn't this much more readable? ;)